### PR TITLE
improvement/in-app-code-samples

### PR DESCRIPTION
**Pull request by dabeeeenster**

### CODE REVIEW

File: frontend/common/code-help/traits/traits-dotnet.js [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
-// Increment a trait
-Trait userTrait = await BulletTrainClient.instance.IncrementTrait("${USER_ID}", "${TRAIT_NAME}", 1);
+static FlagsmithClient _flagsmithClient;
+_flagsmithClient = new("${envId}");
 
-Identity userIdentity = await BulletTrainClient.instance.GetUserIdentity("${USER_ID}");
-if (userIdentity != null) {
-  // Run the code to use user identity i.e. userIdentity.flags or userIdentity.traits
-}
+var Identifier = "delboy@trotterstraders.co.uk";
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-dotnet.js]:
```suggestion
var identifier = "delboy@trotterstraders.co.uk";
```

File: frontend/common/code-help/traits/traits-dotnet.js [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+var traitKey = "car_type";
+var traitValue = "robin_reliant";
+var traitList = new List<Trait> { new Trait(traitKey, traitValue) };
+
+# Sync
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-dotnet.js]: I don't think this is right, is it? Both of these will just return a `Task` and probably the call to `isFeatureEnabled` will actually fail in fact. 

I think you need to add `.Result` to the end of each of these?

File: frontend/common/code-help/traits/traits-flutter.js [hunk 3]
```diff
@@ -1,16 +1,28 @@
...
+//        config: config, 
+//        seeds: <Flag>[
+//            Flag.seed('feature', enabled: true),
+//        ], 
+//        update: false,
+//    );
 
-// Or get specific traits 
-final userTraits = await flagsmithClient.getTraits(user, keys: ['${TRAIT_NAME}', 'other_trait']);
+// This will create a user in the dashboard if they don't already exist
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-flutter.js]: Pretty sure just creating an object in the client won't interact with the API will it? think we should be clear about what point exactly the user is created - this seems a bit misleading to me.

File: frontend/common/code-help/traits/traits-go.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module.exports = (envId, { LIB_NAME, USER_ID, LIB_NAME_JAVA, FEATURE_NAME, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT }, userId) => `var testUser = flagsmith.User{Identifier: "${USER_ID}"}
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-go.js]: Not sure why this documentation is in `frontend/common/code-help/traits/traits-go.js` - there's no reference to traits at all here!

File: frontend/common/code-help/traits/traits-ios.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import Utils from '../../utils/utils';
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-ios.js]: This is Javascript so I'm probably speaking out of turn here but I don't see where we're using the Utils class?

File: frontend/common/code-help/traits/traits-ios.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+func application(_ application: UIApplication,
+ didFinishLaunchingWithOptions launchOptions:
+  [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+
+  Flagsmith.shared.apiKey = "${envId}"
+
+  // This will create a user in the dashboard if they don't already exist
+  // Check for a feature
+  Flagsmith.shared
+  .hasFeatureFlag(withID: "${FEATURE_NAME}", forIdentity: "${USER_ID}") { (result) in
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-ios.js]: Again, no traits here (but in a file named traits-ios.js) 🤔

File: frontend/common/code-help/traits/traits-java.js [hunk 7]
```diff
@@ -1,26 +1,15 @@
...
-    // run the code doesn't depend on user traits
-}
-
-
+module.exports = (envId, { LIB_NAME, USER_ID, TRAIT_NAME, LIB_NAME_JAVA, FEATURE_NAME, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT }, userId) => `${LIB_NAME_JAVA} ${LIB_NAME} = ${LIB_NAME_JAVA}
+.newBuilder()
+.setApiKey("${envId}")
+.build();
+
+// This will create a user in the dashboard if they don't already exist
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-java.js]: I don't think any identities are created until Line 12, right?

File: frontend/common/code-help/traits/traits-node.js [hunk 8]
```diff
@@ -1,6 +1,14 @@
-module.exports = (envId, { LIB_NAME, FEATURE_NAME, TRAIT_NAME, USER_ID, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT, NPM_NODE_CLIENT }, userId) => `import ${LIB_NAME} from "${NPM_NODE_CLIENT}"; // Add this line if you're using ${LIB_NAME} via npm
+module.exports = (envId, { LIB_NAME, FEATURE_NAME, TRAIT_NAME, USER_ID, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT, NPM_NODE_CLIENT }, userId) => `const Flagsmith = require('flagsmith-nodejs');
 
-await ${LIB_NAME}.setTrait("${USER_ID}", ${TRAIT_NAME}, 21)
-const trait = await ${LIB_NAME}.getTrait("${USER_ID}", ${TRAIT_NAME})
+const flagsmith = new Flagsmith(
+    environmentKey: '${envId}'
+);
 
+// This will create a user in the dashboard if they don't already exist
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-node.js]: I don't think the identity is created until line 11, right?

File: frontend/common/code-help/traits/traits-python.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+module.exports = (envId, { LIB_NAME, USER_ID, LIB_NAME_JAVA, FEATURE_NAME, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT }, userId) => `from flagsmith import Flagsmith;
+
+flagsmith = Flagsmith(
+    environment_key = os.environ.get("${envId}")
+)
+
+# This will create a user in the dashboard if they don't already exist
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-python.js]: It's pretty obvious here that we're _not_ doing anything in the 'dashboard' but this comment is still misleading in my opinion.

File: frontend/common/code-help/traits/traits-rest.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+module.exports = (envId, { FEATURE_NAME, FEATURE_FUNCTION, USER_ID, FEATURE_NAME_ALT }) => `# This will create a user in the dashboard if they don't already exist
+
+curl -X "POST" "https://edge.api.flagsmith.com/api/v1/identities/?identifier=winning2"
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-rest.js]: The query param here is redundant (I think the FE just belt and braces it but let's be succinct here and remove it).

File: frontend/common/code-help/traits/traits-rest.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+     -H 'X-Environment-Key: ${envId}'
+     -H 'Content-Type: application/json; charset=utf-8'
+     -d $'{
+  "traits": [
+    {
+      "trait_value": 1000,
+      "trait_key": "gagan_1"
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-rest.js]: Should we really be using Gagan's name as a trait key here? 🤔

File: frontend/common/code-help/traits/traits-rest.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    },
+    {
+      "trait_value": true,
+      "trait_key": "my_other_key"
+    },
+    {
+      "trait_value": "ben2",
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-rest.js]: And your name...!

File: frontend/common/code-help/traits/traits-ruby.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+module.exports = (envId, { FEATURE_NAME, FEATURE_FUNCTION, USER_ID, FEATURE_NAME_ALT }) => `require "flagsmith"
+
+$flagsmith = Flagsmith::Client.new(
+    environment_key: '${envId}'
+)
+
+// This will create a user in the dashboard if they don't already exist.
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-ruby.js]: This is just 100% wrong... No identities will be created when calling the `get_environment_flags` method. Also, there is, again, no mention of traits in `traits-ruby.js`.

**Review by matthewelwell — CHANGES_REQUESTED** [review]

File: frontend/common/code-help/traits/traits-go.js (hunk 4 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-go.js]: Yeah we didnt have that code at the time and I wanted to get this into prod ASAP

File: frontend/common/code-help/traits/traits-flutter.js [hunk 14]
```diff
@@ -1,16 +1,24 @@
-module.exports = (envId, { LIB_NAME, USER_ID, LIB_NAME_JAVA, TRAIT_NAME, FEATURE_NAME, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT }, userId) => `final user = FeatureUser(identifier: '${USER_ID}');
+module.exports = (envId, { LIB_NAME, USER_ID, LIB_NAME_JAVA, TRAIT_NAME, FEATURE_NAME, FEATURE_FUNCTION, FEATURE_NAME_ALT, FEATURE_NAME_ALT_VALUE, NPM_CLIENT }, userId) => `final flagsmithClient = FlagsmithClient(
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-flutter.js]: There's still nothing about traits here?

File: frontend/common/code-help/traits/traits-rest.js (hunk 11 above)
Comment (matthewelwell) [reply frontend/common/code-help/traits/traits-rest.js]: Why was this resolved?

File: frontend/common/code-help/traits/traits-rest.js (hunk 12 above)
Comment (matthewelwell) [reply frontend/common/code-help/traits/traits-rest.js]: And this?

File: frontend/common/code-help/traits/traits-ruby.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+module.exports = (envId, { FEATURE_NAME, FEATURE_FUNCTION, USER_ID, FEATURE_NAME_ALT }) => `require "flagsmith"
+
+$flagsmith = Flagsmith::Client.new(
+    environment_key: '${envId}'
+)
+
+trait_key = params.get(:flagsmith, :trait_key)
```
Comment (matthewelwell) [inline frontend/common/code-help/traits/traits-ruby.js]: ... stop copy pasting!

File: frontend/common/code-help/traits/traits-flutter.js (hunk 14 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-flutter.js]: we need sample code from Foresight

File: frontend/common/code-help/traits/traits-rest.js (hunk 11 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-rest.js]: its just sample data - don't see this as an issue at all.

File: frontend/common/code-help/traits/traits-rest.js (hunk 12 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-rest.js]: Same

File: frontend/common/code-help/traits/traits-ruby.js (hunk 15 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-ruby.js]: Dont follow?

File: frontend/common/code-help/traits/traits-rest.js (hunk 11 above)
Comment (matthewelwell) [reply frontend/common/code-help/traits/traits-rest.js]: I do - it's weird

File: frontend/common/code-help/traits/traits-ruby.js (hunk 15 above)
Comment (matthewelwell) [reply frontend/common/code-help/traits/traits-ruby.js]: This code is completely broken.

Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-ruby.js]: can you fix?

**Comment by 1989shack** [comment]
Full database system Admin and flow the server manager

File: frontend/common/code-help/create-user/create-user-rust.js [hunk 16]
```diff
@@ -9,12 +7,14 @@ let flagsmith = Flagsmith::new(
         .expect("FLAGSMITH_ENVIRONMENT_KEY not found in environment"),
     options,
 );
-// The method below triggers a network request
-let identity_flags = flagsmith.get_identity_flags(identifier, Some(traits)).unwrap();
 
-// Check for a feature
-let show_button = identity_flags.is_feature_enabled("secret_button").unwrap();
+// Identify the user
+let identity_flags = flagsmith.get_identity_flags("${userId}").unwrap();
```
Comment (gagantrivedi) [inline frontend/common/code-help/create-user/create-user-rust.js]: `flagsmith.get_identity_flags("${userId}", None).unwrap();`

File: frontend/common/code-help/traits/traits-rust.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+let options = FlagsmithOptions {..Default::default()};
+let flagsmith = Flagsmith::new(
+    env::var("${envId}")
+        .expect("FLAGSMITH_ENVIRONMENT_KEY not found in environment"),
+    options,
+);
+// The method below triggers a network request
+let identity_flags = flagsmith.get_identity_flags("${USER_ID}", Some(traits)).unwrap();
+`;
+
```
Comment (gagantrivedi) [inline frontend/common/code-help/traits/traits-rust.js]: We can get rid of the initialization here, no?
Adding the traits will look like this:

```
let trait_key = "trait_key1";
let trait_value = "trai_value1";


let traits = vec![Trait {
    trait_key: trait_key.to_string(),
    trait_value: FlagsmithValue {
        value: trait_value.to_string(),
        value_type: FlagsmithValueType::String,
    },
}];
let identity_flags = flagsmith
    .get_identity_flags(identifier, Some(traits))
    .unwrap();



```

**Review by gagantrivedi — CHANGES_REQUESTED** [review]

File: frontend/common/code-help/traits/traits-flutter.js (hunk 14 above)
Comment (dabeeeenster) [reply frontend/common/code-help/traits/traits-flutter.js]: OK this is provided and in the code now

**Review by matthewelwell — APPROVED** [review]

**Pull request merged**